src/main/java/org/opentripplanner/routing/graph/GraphIndex.java

```diff
@@ -434,14 +436,19 @@ public BitSet servicesRunning (LocalDate date) {
         return ret;
     }
 
-    // TODO OTP2 - Add support for includeRealtimeUpdates
+    /**
+     * Returns all the patterns for a specific stop. If includeRealtimeUpdates is set, new patterns
+     * added by realtime updates are added to the collection.
+     */
     public Collection<TripPattern> getPatternsForStop(Stop stop, boolean includeRealtimeUpdates) {
         List<TripPattern> tripPatterns = new ArrayList<>(patternsForStop.get(stop));
-        /*
-        if (includeRealtimeUpdates && graph.timetableSnapshotSource != null) {
-            tripPatterns.addAll(graph.timetableSnapshotSource.getAddedTripPatternsForStop(stop));
+
+        TimetableSnapshot timetableSnapshot = graph.getTimetableSnapshot();
```

Comment:
We just discussed the fact that this is getting a snapshot within a single small method. Snapshots are intended to be held and reused throughout a full request cycle. So methods like this should really be on the snapshot (or applied to a snapshot) rather than applied to the Graph or GraphIndex. The problem is that GraphIndex contains both an index of the essentially immutable data in a Graph, plus a bunch of methods that produce API responses that happen to use that index. But those methods also examine things like realtime data. Really, those API methods should be outside GraphIndex, and they should have a reference to a realtime snapshot, which in turn gives access to an underlying long-lived Graph and GraphIndex.

Reply:
We discussed further - this will work in a loose sense, but it defies some of our assumptions. This method getPatternsForStop is called in many places, various small methods within a GraphQL query. This means a single GraphQL query could fetch different snapshots. It will be easier to reason about this, and to describe behavior to API clients, if we can state concisely that every request always sees exactly one snapshot. There are also performance implications, since the snapshot provider only expects to be called once per request, so may have some slightly time-consuming side effects that amount to committing a transaction.

Reply:
I had a look at this to see if we could refactor to make it better. Gard will try do the following:

Extract a fasade out of the GraphIndex and Graph (lets call it RoutingService). The RoutingService should be the entry point for all access to all logic in the routing java package. Put it on the Routing level. We can let the routing service have a request scope and delegate to the GraphIndex and Graph. Some of the methods in the Graph and Index should be moved where they belong. The RoutingService will have a lazy getter to the TimeTableSnapshot and inject this instance where needed. The APIs should get one RoutingService and use it for the duration of each request. The RoutingService would be a StatefulService.

Reply:
We will try to use the Lombok library @Delegate to make this a true Facade.